datafusion/core/src/physical_plan/row_converter.rs

```diff
@@ -0,0 +1,292 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! [`CardinalityAwareRowConverter`] for converting data to
+//! [`arrow::row`] format.
+
+use arrow::datatypes::DataType;
+use arrow::row::{Row, RowConverter};
+use arrow::row::{Rows, SortField};
+use arrow_array::cast::AsArray;
+use arrow_array::*;
+use datafusion_common::{internal_err, DataFusionError, Result};
+
+/// The threshold of the number of values at which to consider a
+/// [`DictionaryArray`] "high" cardinality.
+///
+/// Since [`RowConverter`] blindly generates a mapping for all values,
+/// regardless of if they appear in the keys, this value is compared
+/// to the length of values.
+///
+/// The assumption is that the number of potential distinct key values
+/// (aka the length of the values array) is a more robust predictor of
+/// being "high" cardinality than the actual number of keys used. The
+/// intuition for this is that other values in the dictionary could be
+/// used in subsequent batches.
+///
+/// While an argument can made for doing something more sophisticated,
+/// this would likely only really make sense if the dictionary
+/// interner itself followed a similar approach, which it did not at
+/// the time of this writing.
+const LOW_CARDINALITY_THRESHOLD: usize = 10;
```

Comment:
I'm curious if this has been determined empirically, a dictionary will almost never have this few values, even if the column itself only contains 2 or 3 distinct values. This might be a very strong argument for removing the dictionary interning from RowConverter...

Reply:
> a dictionary will almost never have this few values, even if the column itself only contains 2 or 3 distinct values. 

I don't understand that statement. 

@JayjeetAtGithub do you have any specific numbers about how this threshold was chosen? @tustvold do you have any suggestions for a different value (or suggestions on how to empirically determine a threshold?)

Reply:
> I don't understand that statement.

Even if an array has very few logically distinct values, there are various mechanisms that will lead the dictionary to contain duplicates and/or values that aren't referenced. As a result almost all dictionaries, certainly any that originate from parquet files, are likely to contain more than 10 values.

> do you have any suggestions for a different value (or suggestions on how to empirically determine a threshold?)

I mean https://github.com/apache/arrow-rs/issues/4811 effectively proposes a value of 0... :sweat_smile:

Reply:
Yeah, the more I think about the costs and benefits of dictionary encoding, the more I am leaning towards trying to remove dictionary encoding entirely in the row format. 

https://github.com/apache/arrow-rs/issues/4811#issuecomment-1717787949